File: src/mcp/__tests__/pkce.test.mjs
import crypto from "node:crypto";
import {
  verifyCodeChallenge,
  isValidCodeVerifier,
  isSupportedChallengeMethod,
} from "../lib/pkce.mjs";

// ---------------------------------------------------------------------------
// Helper: generate a known code_verifier / code_challenge pair for S256
// ---------------------------------------------------------------------------
function makeS256Pair(verifier) {
  const challenge = crypto.createHash("sha256").update(verifier, "ascii").digest("base64url");
  return { verifier, challenge };
}

// ---------------------------------------------------------------------------
// verifyCodeChallenge — S256
// ---------------------------------------------------------------------------
describe("verifyCodeChallenge (S256)", () => {
  test("accepts a valid verifier/challenge pair", () => {
    const { verifier, challenge } = makeS256Pair("dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk");
    expect(verifyCodeChallenge(verifier, challenge, "S256")).toBe(true);
  });

  test("rejects an incorrect verifier", () => {
    const { challenge } = makeS256Pair("correct-verifier-value-that-is-long-enough-here");
    expect(verifyCodeChallenge("wrong-verifier-value-that-is-also-long-enough!", challenge, "S256")).toBe(false);
  });

  test("rejects empty verifier", () => {
    const { challenge } = makeS256Pair("some-verifier-value-that-is-long-enough-12345");
    expect(verifyCodeChallenge("", challenge, "S256")).toBe(false);
  });

  test("rejects empty challenge", () => {
    expect(verifyCodeChallenge("some-verifier", "", "S256")).toBe(false);
  });

  test("rejects null inputs", () => {
    expect(verifyCodeChallenge(null, "challenge", "S256")).toBe(false);
    expect(verifyCodeChallenge("verifier", null, "S256")).toBe(false);
  });

  test("works with RFC 7636 Appendix B test vector", () => {
    // RFC 7636 Appendix B: code_verifier = "dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk"
    // code_challenge (S256) = "E9Melhoa2OwvFrEMTJguCHaoeK1t8URWbuGJSstw-cM"
    const verifier = "dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk";
    const challenge = "E9Melhoa2OwvFrEMTJguCHaoeK1t8URWbuGJSstw-cM";
    expect(verifyCodeChallenge(verifier, challenge, "S256")).toBe(true);
  });
});

// ---------------------------------------------------------------------------
// verifyCodeChallenge — plain
// ---------------------------------------------------------------------------
describe("verifyCodeChallenge (plain)", () => {
  test("accepts matching verifier and challenge", () => {
    const v = "my-plain-code-verifier-that-is-long-enough-here";
    expect(verifyCodeChallenge(v, v, "plain")).toBe(true);
  });

  test("rejects mismatched verifier", () => {
    expect(verifyCodeChallenge("aaa", "bbb", "plain")).toBe(false);
  });

  test("rejects empty strings", () => {
    expect(verifyCodeChallenge("", "", "plain")).toBe(false);
  });
});

// ---------------------------------------------------------------------------
// verifyCodeChallenge — unsupported method
// ---------------------------------------------------------------------------
describe("verifyCodeChallenge (unsupported method)", () => {
  test("rejects unknown method", () => {
    expect(verifyCodeChallenge("v", "c", "S512")).toBe(false);
  });

  test("rejects undefined method", () => {
    expect(verifyCodeChallenge("v", "c", undefined)).toBe(false);
  });
});

// ---------------------------------------------------------------------------
// isValidCodeVerifier
// ---------------------------------------------------------------------------
describe("isValidCodeVerifier", () => {
  test("accepts 43-character verifier", () => {
    expect(isValidCodeVerifier("a".repeat(43))).toBe(true);
  });

  test("accepts 128-character verifier", () => {
    expect(isValidCodeVerifier("B".repeat(128))).toBe(true);
  });

  test("accepts verifier with all unreserved characters", () => {
    expect(isValidCodeVerifier("ABCDEFghij0123456789-._~" + "x".repeat(19))).toBe(true);
  });

  test("rejects verifier shorter than 43 characters", () => {
    expect(isValidCodeVerifier("a".repeat(42))).toBe(false);
  });

  test("rejects verifier longer than 128 characters", () => {
    expect(isValidCodeVerifier("a".repeat(129))).toBe(false);
  });

  test("rejects verifier with invalid characters", () => {
    expect(isValidCodeVerifier("a".repeat(42) + " ")).toBe(false);
    expect(isValidCodeVerifier("a".repeat(42) + "+")).toBe(false);
    expect(isValidCodeVerifier("a".repeat(42) + "/")).toBe(false);
    expect(isValidCodeVerifier("a".repeat(42) + "=")).toBe(false);
  });

  test("rejects non-string", () => {
    expect(isValidCodeVerifier(123)).toBe(false);
    expect(isValidCodeVerifier(null)).toBe(false);
    expect(isValidCodeVerifier(undefined)).toBe(false);
  });

  test("rejects empty string", () => {
    expect(isValidCodeVerifier("")).toBe(false);
  });
});

// ---------------------------------------------------------------------------
// isSupportedChallengeMethod
// ---------------------------------------------------------------------------
describe("isSupportedChallengeMethod", () => {
  test("accepts S256", () => {
    expect(isSupportedChallengeMethod("S256")).toBe(true);
  });

  test("accepts plain", () => {
    expect(isSupportedChallengeMethod("plain")).toBe(true);
  });

  test("rejects S512", () => {
    expect(isSupportedChallengeMethod("S512")).toBe(false);
  });

  test("rejects empty string", () => {
    expect(isSupportedChallengeMethod("")).toBe(false);
  });

  test("rejects undefined", () => {
    expect(isSupportedChallengeMethod(undefined)).toBe(false);
  });
});

File: src/server/gkill/mcp/pkce_test.go
package mcp

// 旧 src/mcp/__tests__/pkce.test.mjs の移植。

import (
	"crypto/sha256"
	"encoding/base64"
	"strings"
	"testing"
)

// makeS256Pair は既知の code_verifier / code_challenge の対を作る。
func makeS256Pair(verifier string) (string, string) {
	sum := sha256.Sum256([]byte(verifier))
	return verifier, base64.RawURLEncoding.EncodeToString(sum[:])
}

func TestVerifyCodeChallengeS256(t *testing.T) {
	t.Run("accepts a valid verifier/challenge pair", func(t *testing.T) {
		verifier, challenge := makeS256Pair("dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk")
		expectTrue(t, VerifyCodeChallenge(verifier, challenge, "S256"), "must verify")
	})
	t.Run("rejects an incorrect verifier", func(t *testing.T) {
		_, challenge := makeS256Pair("correct-verifier-value-that-is-long-enough-here")
		expectTrue(t, !VerifyCodeChallenge("wrong-verifier-value-that-is-also-long-enough!", challenge, "S256"), "must reject")
	})
	t.Run("rejects empty verifier", func(t *testing.T) {
		_, challenge := makeS256Pair("some-verifier-value-that-is-long-enough-12345")
		expectTrue(t, !VerifyCodeChallenge("", challenge, "S256"), "must reject")
	})
	t.Run("rejects empty challenge", func(t *testing.T) {
		expectTrue(t, !VerifyCodeChallenge("some-verifier", "", "S256"), "must reject")
	})
	t.Run("rejects null inputs", func(t *testing.T) {
		expectTrue(t, !VerifyCodeChallenge("", "challenge", "S256"), "null verifier")
		expectTrue(t, !VerifyCodeChallenge("verifier", "", "S256"), "null challenge")
	})
	t.Run("works with RFC 7636 Appendix B test vector", func(t *testing.T) {
		// RFC 7636 Appendix B: code_verifier = "dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk"
		// code_challenge (S256) = "E9Melhoa2OwvFrEMTJguCHaoeK1t8URWbuGJSstw-cM"
		verifier := "dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk"
		challenge := "E9Melhoa2OwvFrEMTJguCHaoeK1t8URWbuGJSstw-cM"
		expectTrue(t, VerifyCodeChallenge(verifier, challenge, "S256"), "RFC vector must verify")
	})
}

// verifyCodeChallenge — plain は廃止 (S256 必須)。plain は verifier==challenge のため
// 中間者が challenge を書き換えるだけで無効化でき、OAuth 2.1 / MCP でも禁止されている。
func TestVerifyCodeChallengePlainRejected(t *testing.T) {
	t.Run("rejects plain even when verifier equals challenge", func(t *testing.T) {
		v := "my-plain-code-verifier-that-is-long-enough-here"
		expectTrue(t, !VerifyCodeChallenge(v, v, "plain"), "plain must be rejected")
	})
}

func TestVerifyCodeChallengeUnsupportedMethod(t *testing.T) {
	t.Run("rejects unknown method", func(t *testing.T) {
		expectTrue(t, !VerifyCodeChallenge("v", "c", "S512"), "S512")
	})
	t.Run("rejects undefined method", func(t *testing.T) {
		expectTrue(t, !VerifyCodeChallenge("v", "c", ""), "undefined")
	})
}

func TestIsValidCodeVerifier(t *testing.T) {
	t.Run("accepts 43-character verifier", func(t *testing.T) {
		expectTrue(t, IsValidCodeVerifier(strings.Repeat("a", 43)), "43")
	})
	t.Run("accepts 128-character verifier", func(t *testing.T) {
		expectTrue(t, IsValidCodeVerifier(strings.Repeat("B", 128)), "128")
	})
	t.Run("accepts verifier with all unreserved characters", func(t *testing.T) {
		expectTrue(t, IsValidCodeVerifier("ABCDEFghij0123456789-._~"+strings.Repeat("x", 19)), "unreserved")
	})
	t.Run("rejects verifier shorter than 43 characters", func(t *testing.T) {
		expectTrue(t, !IsValidCodeVerifier(strings.Repeat("a", 42)), "42")
	})
	t.Run("rejects verifier longer than 128 characters", func(t *testing.T) {
		expectTrue(t, !IsValidCodeVerifier(strings.Repeat("a", 129)), "129")
	})
	t.Run("rejects verifier with invalid characters", func(t *testing.T) {
		expectTrue(t, !IsValidCodeVerifier(strings.Repeat("a", 42)+" "), "space")
		expectTrue(t, !IsValidCodeVerifier(strings.Repeat("a", 42)+"+"), "plus")
		expectTrue(t, !IsValidCodeVerifier(strings.Repeat("a", 42)+"/"), "slash")
		expectTrue(t, !IsValidCodeVerifier(strings.Repeat("a", 42)+"="), "equals")
	})
	t.Run("rejects non-string", func(t *testing.T) {
		expectTrue(t, !IsValidCodeVerifier(123), "number")
		expectTrue(t, !IsValidCodeVerifier(nil), "null")
	})
	t.Run("rejects empty string", func(t *testing.T) {
		expectTrue(t, !IsValidCodeVerifier(""), "empty")
	})
}

func TestIsSupportedChallengeMethod(t *testing.T) {
	t.Run("accepts S256", func(t *testing.T) {
		expectTrue(t, IsSupportedChallengeMethod("S256"), "S256")
	})
	t.Run("rejects plain (S256 only)", func(t *testing.T) {
		expectTrue(t, !IsSupportedChallengeMethod("plain"), "plain")
	})
	t.Run("rejects S512", func(t *testing.T) {
		expectTrue(t, !IsSupportedChallengeMethod("S512"), "S512")
	})
	t.Run("rejects empty string", func(t *testing.T) {
		expectTrue(t, !IsSupportedChallengeMethod(""), "empty")
	})
	t.Run("rejects undefined", func(t *testing.T) {
		expectTrue(t, !IsSupportedChallengeMethod(""), "undefined")
	})
}
